### PR TITLE
Restrict `readFile` context to references that appear in the string

### PR DESCRIPTION
**Pull request by ncfavier**
When calling `builtins.readFile` on a store path, the references of that path are currently added to the resulting string's context (https://github.com/NixOS/nix/pull/1643).

This change makes those references the *possible* context of the string, but filters them to keep only the references whose hash actually appears in the string, similarly to what is done for determining the runtime references of a path.

Fixes https://github.com/NixOS/nix/issues/6647 and probably https://github.com/NixOS/nixpkgs/issues/175125 (not sure how to test that one) while keeping the `readfile-context` test successful.

~I wasn't sure how to generalise `dumpPath(path, sink)` and `sink << string` so I gave in to my primal Haskeller instincts and used a lambda. We could make `scanForReferences` accept a source, but then we'd need to turn `λ sink. dumpPath(path, sink)` into a source and the simplest way to do this [seems](https://github.com/NixOS/nix/blob/499e99d099ec513478a2d3120b2af3a16d9ae49d/src/libutil/serialise.cc#L245-L292) to involves coroutines...~

### CODE REVIEW

**Comment by Ericson2314** [comment]
This is a nice idea, thanks for looking into this!

File: src/libstore/references.cc
Comment (Ericson2314) [inline src/libstore/references.cc:94]:
```suggestion
    Source & source,
```
I feel like this should make it less awkward? Note that there are various source to sink and sink to source functions.

Comment (ncfavier) [reply src/libstore/references.cc:94]: Yes, I noticed `sinkToSource` (see OP), but I feel like it's a shame to use coroutines for something so simple...

Comment (Ericson2314) [reply src/libstore/references.cc:94]: Actually, what we really want is a `PathRefScanSink`

Comment (ncfavier) [reply src/libstore/references.cc:94]: That probably makes more sense

Comment (ncfavier) [reply src/libstore/references.cc:94]: Ugh, this is apparently too much for C++. Tried `PathRefScanSink : public RefScanSink`, but it doesn't seem possible to do the preprocessing (for each path, add its hash to `hashes` and insert it into the `backMap`) *before* calling the parent constructor. I'd like to avoid having an inner `RefScanSink * sink;` member and wiring it up explicitly...

I guess I'll have to make a "smart constructor" function that does the preprocessing and calls the constructor.

Comment (Ericson2314) [reply src/libstore/references.cc]: @ncfavier Here's a trick, you can make static method (or just free-floating function) which the constructor calls!

Comment (ncfavier) [reply src/libstore/references.cc]: But this needs to happen in the member initialiser list, so we can't feed one output of the function (hashes) to the base constructor and the other output (backMap) to initialise a member...

Comment (Ericson2314) [reply src/libstore/references.cc]: @ncfavier I was thinking about this, see https://stackoverflow.com/questions/4162021/is-it-ok-to-call-a-function-in-constructor-initializer-list I think we can make a blank back map, then make the other one set by mutating the first one.

Comment (Ericson2314) [reply src/libstore/references.cc]: @ncfavier I replied on matrix with details.

Comment (ncfavier) [reply src/libstore/references.cc]: Made the constructor private and inlined `scanStringForReferences` as requested

Comment (Ericson2314) [reply src/libstore/references.cc]: Thanks!

**Review by Ericson2314 — APPROVED** [review]
This change is looking quite good. It is the best of both worlds between not just dropping the references but also having less of a breaking change than 2.3.

Thank you very much @ncfavier!

File: src/libstore/references.hh
Comment (Ericson2314) [inline src/libstore/references.hh:37]:
```suggestion
    PathRefScanSink(StringSet && hashes, std::map<std::string, StorePath> && backMap)
        : RefScanSink(std::move(hashes))
        , backMap(std::move(backMap))
    { }
```

Comment (Ericson2314) [reply src/libstore/references.hh:37]: I would also move it out of the header into the implementation file with
```suggestion
    PathRefScanSink(StringSet && hashes, std::map<std::string, StorePath> backMap);
```
in just the header,

That leaks the least amount of what is basically implementation details given the "smart constructor".

File: src/libstore/references.cc
Comment (Ericson2314) [inline src/libstore/references.cc:70]: if you do opt to merely declare in the header as I suggested
```suggestion
PathRefScanSink::PathRefScanSink(StringSet && hashes, std::map<std::string, StorePath> && backMap)
    : RefScanSink(std::move(hashes))
    , backMap(std::move(backMap))
{ }
    
PathRefScanSink PathRefScanSink::fromPaths(const StorePathSet & refs)
```

File: src/libexpr/primops.cc
Comment (Ericson2314) [inline src/libexpr/primops.cc:1546]:
```suggestion
        // Re-scan refs to filter down to just the ones that actually occur in the file.
        PathRefScanSink refsSink = PathRefScanSink::fromPaths(refs);
```

**Review by Ericson2314 — APPROVED** [review]

File: src/libexpr/primops.cc
Comment (edolstra) [inline src/libexpr/primops.cc:1547]:
```suggestion
        auto refsSink = PathRefScanSink::fromPaths(refs);
```

**Review by thufschmitt — APPROVED** [review]
Thanks!

@ncfavier I've pushed some release notes and a test. I'll let you confirm that they look all right

File: tests/readfile-context.nix
Comment (ncfavier) [inline tests/readfile-context.nix:23]: Apparently this succeeds even if there's a context.

Comment (thufschmitt) [reply tests/readfile-context.nix:23]: How did you see that? I tried replacing `discardableDependent` by `dependent` and the evaluation fails because `the string '/tmp/nix-test/tests/readfile-context/store/dk4iy6vpksymyjpm1wl0646hh5p3vm3c-simple/hello' is not allowed to refer to a store path `.

I also tried with `builtins.substring 0 0 (builtins.readFile dependent) + "foo"` just to make sure that I had the same error with a valid `name`, and I indeed have it

Comment (ncfavier) [reply tests/readfile-context.nix:23]: Ah, I see, the problem is that `discardableDependent` has no references at all so the test succeeds even without this patch.

What we should do to test this properly is have a derivation that writes *two* files, one with a reference and one without, and read the one without a reference.

Comment (thufschmitt) [reply tests/readfile-context.nix:23]: Oh yes, of course! I assumed that `discardableDepend` would refer to itself, but it isn't by default

Comment (ncfavier) [reply tests/readfile-context.nix]: Updated the test.

**Review by thufschmitt — APPROVED** [review]

**Comment by thufschmitt** [comment]
Just waiting for the CI to get green, and I'll merge

**Comment by ncfavier** [comment]
It's green now!

**Comment by thufschmitt** [comment]
> It's green now!

Yeah sorry, I kinda new when writing my message that I would forget about it :unamused:

**Pull request merged**
